webknossos-datastore/app/com/scalableminds/webknossos/datastore/tracings/skeleton/TreeValidator.scala

```diff
@@ -102,15 +101,66 @@ object TreeValidator {
       }
     }
 
+  def checkAllNodesUsedInCommentsExist(trees: Seq[Tree], comments: Seq[Comment]): Box[Unit] = {
+    checkAllNodesUsedExist(trees, comments.map(_.nodeId).distinct) match {
+      case Some(nodesOnlyInComments) => Failure(s"Some comments refer to non-existent nodes. comments: ${nodesOnlyInComments.mkString(", ")}")
+      case None => Full(())
+    }
+  }
+
+  def checkAllNodesUsedInBranchPointsExist(trees: Seq[Tree], branchPoints: Seq[BranchPoint]): Box[Unit] = {
+    checkAllNodesUsedExist(trees, branchPoints.map(_.nodeId).distinct) match {
+      case Some(nodesOnlyInBranchPoints) => Failure(s"Some branchPoints refer to non-existent nodes. branchPoints: ${nodesOnlyInBranchPoints.mkString(", ")}")
+      case None => Full(())
+    }
+  }
+
+  def checkNoDuplicateTreeGroupIds(treeGroups: Seq[TreeGroup]): Box[Unit] = {
+    val treeGroupIds = getAllTreeGroupIds(treeGroups, Seq[Int]())
+    val distinctTreeGroupIds = treeGroupIds.distinct
+    if (treeGroupIds.size == distinctTreeGroupIds.size) {
+      Full(())
+    } else {
+      Failure(s"Duplicate treeGroupIds: ${treeGroupIds.diff(distinctTreeGroupIds).mkString(", ")}")
+    }
+  }
 
+  def checkAllTreeGroupIdsUsedExist(trees: Seq[Tree], treeGroups: Seq[TreeGroup]): Box[Unit] = {
+    val existingTreeGroups = getAllTreeGroupIds(treeGroups, Seq[Int]())
+    val treeGroupsInTrees = trees.flatMap(_.groupId).distinct
+
+    val treeGroupsOnlyInTrees = treeGroupsInTrees.diff(existingTreeGroups)
+    if (treeGroupsOnlyInTrees.isEmpty) {
+      Full(())
+    } else {
+      Failure(s"Some treeGroups used in trees don't exist. treeGroups: ${treeGroupsOnlyInTrees.mkString(", ")}")
+    }
+  }
 
   private def foldOverTrees(trees: Seq[Tree])(block: Tree => Box[Unit]) = {
-    trees.foldLeft[Box[Unit]](Full(())){
+    trees.foldLeft[Box[Unit]](Full(())) {
       case (Full(()), tree) =>
         block(tree)
       case (f, _) =>
         f
     }
   }
 
+  private def getAllTreeGroupIds(treeGroups: Seq[TreeGroup], ids: Seq[Int]): Seq[Int] = {
+    treeGroups match {
+      case (head :: tail) => getAllTreeGroupIds(tail ++ head.children, head.groupId +: ids)
+      case _ => ids
+    }
+  }
+
+  private def checkAllNodesUsedExist(trees: Seq[Tree], usedNodes: Seq[Int]) = {
```

Comment:
As this is called `check…`, please include the actual check with the `Failure` message here. It justs needs the name of the nodes for the message.